File: sdks/go/transport.go
package greplog

import (
	"encoding/binary"
	"log"
	"net"
	"strconv"
	"sync"
	"time"
)

const (
	defaultSocketPath     = ".greplog/greplog.sock"
	defaultTCPHost        = "127.0.0.1"
	defaultTCPPort        = 4318
	maxPendingFrames      = 1000
	defaultReconnectDelay = 5 * time.Second
)

type transport struct {
	connMu    sync.Mutex
	conn      net.Conn
	sockPath  string
	tcpHost   string
	tcpPort   int
	closed    bool
	writeCh   chan []byte
	pending   [][]byte
	pendingMu sync.Mutex
}

func newTransport(sockPath, tcpHost string, tcpPort int, reconnectDelay time.Duration) *transport {
	if sockPath == "" {
		sockPath = defaultSocketPath
	}
	if tcpHost == "" {
		tcpHost = defaultTCPHost
	}
	if tcpPort == 0 {
		tcpPort = defaultTCPPort
	}
	if reconnectDelay <= 0 {
		reconnectDelay = defaultReconnectDelay
	}
	t := &transport{
		sockPath: sockPath,
		tcpHost:  tcpHost,
		tcpPort:  tcpPort,
		writeCh:  make(chan []byte, 256),
	}
	go t.run(reconnectDelay)
	return t
}

func (t *transport) send(data []byte) {
	if t == nil {
		return
	}
	select {
	case t.writeCh <- data:
	default:
	}
}

func (t *transport) shutdown() {
	t.connMu.Lock()
	t.closed = true
	if t.conn != nil {
		t.conn.Close()
	}
	t.connMu.Unlock()
}

func (t *transport) trySendPending() {
	if t.closed {
		return
	}

	t.pendingMu.Lock()
	if len(t.pending) == 0 {
		t.pendingMu.Unlock()
		return
	}
	frames := t.pending
	t.pending = nil
	t.pendingMu.Unlock()

	t.connMu.Lock()
	conn := t.tryConnect()
	t.connMu.Unlock()

	if conn == nil {
		t.pendingMu.Lock()
		t.pending = append(t.pending, frames...)
		if len(t.pending) > maxPendingFrames {
			t.pending = t.pending[len(t.pending)-maxPendingFrames:]
		}
		t.pendingMu.Unlock()
		printWarning()
		return
	}

	for i, frame := range frames {
		if err := writeFrame(conn, frame); err != nil {
			t.connMu.Lock()
			if t.conn == conn {
				t.conn.Close()
				t.conn = nil
			}
			t.connMu.Unlock()
			t.pendingMu.Lock()
			t.pending = append(t.pending, frames[i:]...)
			if len(t.pending) > maxPendingFrames {
				t.pending = t.pending[len(t.pending)-maxPendingFrames:]
			}
			t.pendingMu.Unlock()
			printWarning()
			return
		}
	}
}

func (t *transport) tryConnect() net.Conn {
	if t.closed {
		return nil
	}
	if t.conn != nil {
		_, err := t.conn.Write(nil)
		if err == nil {
			return t.conn
		}
		t.conn.Close()
		t.conn = nil
	}
	conn := t.dial()
	t.conn = conn
	if conn != nil {
		t.flushPending(conn)
	}
	return conn
}

func (t *transport) dial() net.Conn {
	conn := t.dialUDS()
	if conn != nil {
		return conn
	}
	return t.dialTCP()
}

func (t *transport) dialUDS() net.Conn {
	conn, err := net.DialTimeout("unix", t.sockPath, 5*time.Second)
	if err != nil {
		return nil
	}
	return conn
}

func (t *transport) dialTCP() net.Conn {
	addr := net.JoinHostPort(t.tcpHost, strconv.Itoa(t.tcpPort))
	conn, err := net.DialTimeout("tcp", addr, 5*time.Second)
	if err != nil {
		return nil
	}
	return conn
}

func (t *transport) flushPending(conn net.Conn) {
	t.pendingMu.Lock()
	pending := t.pending
	t.pending = nil
	t.pendingMu.Unlock()

	for _, frame := range pending {
		if err := writeFrame(conn, frame); err != nil {
			t.pendingMu.Lock()
			t.pending = append(t.pending, frame)
			t.pendingMu.Unlock()
			return
		}
	}
}

func (t *transport) pend(frame []byte) {
	t.pendingMu.Lock()
	t.pending = append(t.pending, frame)
	if len(t.pending) > maxPendingFrames {
		t.pending = t.pending[1:]
	}
	t.pendingMu.Unlock()
}

func (t *transport) run(reconnectDelay time.Duration) {
	ticker := time.NewTicker(reconnectDelay)
	defer ticker.Stop()

	for {
		if t.closed {
			return
		}

		select {
		case frame, ok := <-t.writeCh:
			if !ok {
				return
			}
			t.pend(frame)
		case <-ticker.C:
		}

		t.trySendPending()
	}
}

func writeFrame(conn net.Conn, data []byte) error {
	header := make([]byte, 4)
	binary.LittleEndian.PutUint32(header, uint32(len(data)))
	if _, err := conn.Write(header); err != nil {
		return err
	}
	if _, err := conn.Write(data); err != nil {
		return err
	}
	return nil
}

var warningPrinted bool
var warningMu sync.Mutex

func printWarning() {
	warningMu.Lock()
	defer warningMu.Unlock()
	if warningPrinted {
		return
	}
	warningPrinted = true
	log.Println("[Greplog] Agent not found. Run 'greplog dev' to capture logs.")
}

File: sdks/rust/src/transport.rs
use std::io::Write;
use std::net::TcpStream;
use std::os::unix::net::UnixStream;
use std::sync::mpsc::{self, Receiver, Sender};
use std::thread;
use std::time::Duration;

const DEFAULT_SOCKET_PATH: &str = ".greplog/greplog.sock";
const DEFAULT_TCP_HOST: &str = "127.0.0.1";
const DEFAULT_TCP_PORT: u16 = 4318;
const DEFAULT_RECONNECT_DELAY: Duration = Duration::from_secs(5);

#[allow(dead_code)]
enum Cmd {
    Send(Vec<u8>),
    #[allow(dead_code)]
    Shutdown,
}

pub struct Transport {
    tx: Sender<Cmd>,
}

impl Transport {
    pub fn new(
        socket_path: Option<&str>,
        tcp_host: Option<&str>,
        tcp_port: Option<u16>,
        reconnect_delay: Option<Duration>,
    ) -> Self {
        let (tx, rx) = mpsc::channel::<Cmd>();

        let sock = socket_path.unwrap_or(DEFAULT_SOCKET_PATH).to_string();
        let host = tcp_host.unwrap_or(DEFAULT_TCP_HOST).to_string();
        let port = tcp_port.unwrap_or(DEFAULT_TCP_PORT);
        let delay = reconnect_delay.unwrap_or(DEFAULT_RECONNECT_DELAY);

        thread::spawn(move || run(rx, &sock, &host, port, delay));

        Transport { tx }
    }

    pub fn send(&self, data: Vec<u8>) {
        let _ = self.tx.send(Cmd::Send(data));
    }

    #[allow(dead_code)]
    pub fn shutdown(&self) {
        let _ = self.tx.send(Cmd::Shutdown);
    }
}

fn run(rx: Receiver<Cmd>, sock_path: &str, tcp_host: &str, tcp_port: u16, delay: Duration) {
    let mut pending: Vec<Vec<u8>> = Vec::new();
    let mut warned = false;

    loop {
        if !pending.is_empty() {
            if let Some(mut conn) = connect(sock_path, tcp_host, tcp_port) {
                pending.retain(|frame| write_frame(&mut conn, frame).is_err());
            } else if !warned {
                warned = true;
                eprintln!("[greplog] Agent not found. Run 'greplog dev' to capture logs.");
            }
        }

        match rx.recv_timeout(delay) {
            Ok(Cmd::Send(data)) => {
                pending.push(data);
            }
            Ok(Cmd::Shutdown) => break,
            Err(mpsc::RecvTimeoutError::Timeout) => {}
            Err(mpsc::RecvTimeoutError::Disconnected) => break,
        }
    }
}

fn connect(sock_path: &str, tcp_host: &str, tcp_port: u16) -> Option<Box<dyn Write + Send>> {
    if let Ok(stream) = UnixStream::connect(sock_path) {
        stream.set_write_timeout(Some(Duration::from_secs(5))).ok();
        Some(Box::new(stream))
    } else if let Ok(stream) = TcpStream::connect((tcp_host, tcp_port)) {
        stream.set_write_timeout(Some(Duration::from_secs(5))).ok();
        Some(Box::new(stream))
    } else {
        None
    }
}

fn write_frame(conn: &mut dyn Write, data: &[u8]) -> Result<(), ()> {
    let len = data.len() as u32;
    let header = len.to_le_bytes();
    conn.write_all(&header).map_err(|_| ())?;
    conn.write_all(data).map_err(|_| ())?;
    conn.flush().map_err(|_| ())?;
    Ok(())
}
